### PR TITLE
[Merged by Bors] - feat: port linarith

### PR DESCRIPTION
**Pull request by kim-em**
TODO:
* Before merge
- [x] fix a bug in linarith in mathlib3 I just found ...
- [x] depends on: #519 
- [x] style lint
- [x] docs
- [X] move theory stubs to a separate PR, for easier tracking: #733 
- [x] failing to parse the `LinarithConfig` option

* Before or after merge?
- [ ] Implement the `removeNe` preprocessor.
- [ ] Add support for restricting to a single type. How to store a `Type` in `LinarithConfig`?

* After merge
- [ ] Teach `norm_num` to solve `example [LinearOrderedRing α] : (0 : α) < 37 := by norm_num`.
- [ ] Port `zify_proof` (plumbing for `zify`), and add the `natToInt` preprocessor.
  Mostly done now, but see #741 before all the tests will work.
- [ ] Port `cancel_denoms` tactic, and add the `cancelDenoms` preprocessor.
- [ ] Add the `nlinarith` preprocessor and front-end syntax.

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#519~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
I've just brought this up to date with `master`, but some typeclass instances got lost in the process. I'll try to fix these tonight.

File: Mathlib/Tactic/Linarith/Datatypes.lean [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+If it is, it returns the comparison along with `t`.
+-/
+def parseCompAndExpr (e : Expr) : MetaM (Ineq × Expr) := do
+  let e ← instantiateMVars e
+  match e.getAppFnArgs with
+  | (``LT.lt, #[_, _, e, z]) => if z.zero? then return (Ineq.lt, e) else throwNotZero z
+  | (``LE.le, #[_, _, e, z]) => if z.zero? then return (Ineq.le, e) else throwNotZero z
+  | (``Eq, #[_, e, z]) => if z.zero? then return (Ineq.eq, e) else throwNotZero z
+  | _ => throwError "invalid comparison: {e}"
+  where throwNotZero (z : Expr) := throwError "invalid comparison, rhs not zero: {z}"
```
Comment (semorrison) [inline Mathlib/Tactic/Linarith/Datatypes.lean:386]:
```suggestion
  where /-- helper function for error message -/
  throwNotZero (z : Expr) := throwError "invalid comparison, rhs not zero: {z}"
```

**Comment by digama0** [comment]
There are still some style things to clean up, and ideally it would use `Qq` more than it does, but this is good enough as a base for future modifications and it's passing the tests.

bors merge

**Comment by semorrison** [comment]
bors p=10

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**